src/ucp/core/ucp_context.h

```diff
@@ -497,17 +497,16 @@ ucp_memory_detect_internal(ucp_context_h context, const void *address,
     }
 
     status = ucs_memtype_cache_lookup(address, length, mem_info);
-    if (ucs_likely(status == UCS_ERR_NO_ELEM)) {
-        goto out_host_mem;
-    }
-
-    ucs_assertv((status == UCS_OK) || (status == UCS_ERR_UNSUPPORTED),
-                "%s (%d)", ucs_status_string(status), status);
-
-    if ((status == UCS_ERR_UNSUPPORTED) ||
-        (mem_info->type == UCS_MEMORY_TYPE_UNKNOWN) ||
-        (mem_info->sys_dev == UCS_SYS_DEVICE_ID_UNKNOWN)) {
+    if (ucs_unlikely((status == UCS_ERR_UNSUPPORTED) ||
```

Comment:
1. pls simplify, i think we can avoid "goto" now
2. host memory is the fast path and should be checked first

Reply:
> host memory is the fast path and should be checked first

done

> pls simplify, i think we can avoid "goto" now

can't remove `goto` - two places use it - and can't merge them into single `if` statement, because `context->num_mem_type_detect_mds == 0` must be checked before calling `ucs_memtype_cache_lookup()`